doc/adr/0004-git-branching-strategy.md

```diff
@@ -0,0 +1,86 @@
+# ADR-0004. Git Branching Strategy
+
+Date: 12/11/2019
+
+## Status
+
+Accepted
+
+## Context
+
+A team utilising git repositories should have a consistent set of branching
+strategies. This will enable encoding relevant information into a branch name
+for improved efficiency. Each component of the branch name will then be able to
+be utilised in other areas of automation.
+
+## Decision
+
+Each branch _**will**_ have it's own deployed environment to use for testing the
+developed code. There will be a 1 to 1 relationship between deployed
+environments and repository brnaches. The only exception to this 1 to 1
+relationship will be the _**master**_ branch. The _**master**_ branch may be used to
+back any environment that needs to be live like.
+
+Each branch name will be broken down into 3 components. An optional category,
+default-environment-reference reference and a brief summary. The approved format
+is shown below.
+
+  `<category>/<default-environment-reference>-<summary>`
+  OR
+  `<default-environment-reference>-<summary>`
```

Comment:
I think we should use `jira-id` rather than `default-environment-reference` here, and add another variant for personal branches, in order to separate out the way we manage environments from our branching strategy a little. I think it also makes this easier to use as a reference later on.

For a different discussion, we'll probably want to deploy environments based on the full branch name, rather than a portion, because we could see multiple branches for the same issue. Decoupling the deployment strategy from branching strategy, but making sure they're compatible feels cleaner.

Reply:
jira-id makes sense, I was trying to decouple us from specific tools. But we can always update this later if we switch.

deploying off the full branch name will raise a whole heap of other issues. for instance to properly namespace resources in azure the name attribute will need to have the env-name prefixed onto it. we will quickly run into issues, some resources don't like certain characters, some only want lower case letters. There is almost certainly a character limit which, `infra-32-creating-a-new-thing` will seriously eat into. (that short name is 29 characters.) VM names for example have a limit of 64 characters.

Reply:
:+1: good point about the tooling. Maybe `issue-id` to keep it generic?

Sounds like there's a good discussion to be had around (auto)deployment models later on. Exciting times 🥳

Reply:
I like `issue-id`